### ISSUE TITLE
Commands to reproduce the results of Table 4 

### ISSUE THREAD

**Issue by dorost1234** [shmsw25/Channel-LM-Prompting#3]
Hi,
I am trying to reproduce the Table 4. 
- Could you tell me please if the results are on dev or test set? 
- Did you run each model for multiple learning rates and then report the test results for the model obtaining the best on the dev set? did you do this for all methods? seems lr for prompt-tuning is finetuned between {0.1, 0.01, 0.001} and for the other method is set to 1e-5? (page 13) 
- Do I need to change any hyper-parameters in the scripts? (this is said global step is set to 100 in the paper, but I could not find it within parameters to set it).
- this is written you have run each model for 80 times, shall I do it with 80 different seeds? 
- you mentioned "Note that, by default, we use the checkpoint that is trained for 100 steps.", I was wondering why not choosing the steps that performing the best on the validation set?
- Could you share with me the commands you used for Table 4? Here please find the sample commands I am planning to use, not sure if this can be all I need to set:

```
// Direct with Head 
python main.py     --task trec     --split dev     --data_dir data     --out_dir out     --gpt2 gpt2         --method direct --do_train --head_tune

//Direct Transform 
python main.py     --task trec     --split dev     --data_dir data     --out_dir out     --gpt2 gpt2         --method direct --do_train --transform_tune 

//Direct prompt
python main.py     --task trec     --split dev     --data_dir data     --out_dir out     --gpt2 gpt2         --method direct --do_train --prompt_tune

//Channel model with prompt 
python main.py     --task trec     --split dev     --data_dir data     --out_dir out     --gpt2 gpt2         --method channel --do_train --prompt_tune
```

Shall I run all above commands with multiple lrs and then report the test set for the model obtaining the best average accuracy? would this be the same hyper-parameters/setup as you did for Table 4? really appreciate letting me know if I am missing some setups for each model. 

thanks

**Comment by shmsw25** [comment]
> Could you tell me please if the results are on dev or test set?

All results are on the test set.

> Did you run each model for multiple learning rates and then report the test results for the model obtaining the best on the dev set? did you do this for all methods? seems lr for prompt-tuning is finetuned between {0.1, 0.01, 0.001} and for the other method is set to 1e-5? (page 13)

Details are provided in Section 5.2 of the paper. To summarize: we did not use the dev set, and followed previous papers for hyperparams. Learning rate was chosen from {0.1, 0.01, 0.001} based on the training loss, for all methods except standard finetuning --- this is because the methods that finetunes a small subset of parameters are known to require larger batch sizes. Only for the standard finetuning, we used 1e-05 because that is the default hyperparameter of many of the HF example codes.

> Do I need to change any hyper-parameters in the scripts? (this is said global step is set to 100 in the paper, but I could not find it within parameters to set it).

global step is already set to 100 - no need to specify it in the command line. The only hyperparam you need to change is learning rate - you can follow the values reported in the Appendix of the paper.

> this is written you have run each model for 80 times, shall I do it with 80 different seeds?

Yes, I used {13,21,42,87,100} for data seeds, {1,10,100,1000} for training seeds. So it's 5 data seeds * 4 training seeds * 4 templates = 80 runs.

> you mentioned "Note that, by default, we use the checkpoint that is trained for 100 steps.", I was wondering why not choosing the one performing the best on the validation set?

The motivation is described in Section 5.2 of the paper. In short, in order to use the validation data for 16-shot learning, 16 datapoints have to be split into the training set and the dev set. In other words, the training data has to be smaller than 16. Nonetheless, previous work (e.g. [Perez et al 2021](https://arxiv.org/abs/2105.11447)) has found that choosing hyperparams based on cross-validation is not much better than random hyperparams. Therefore, we thought that using more datapoints for training would be much better than having the validation set, especially as having more training examples is so crucial for training the model (as shown in Figure 4).

> Could you share with me the commands you used for Table 4? 

Please see the [instructions in README](https://github.com/shmsw25/Channel-LM-Prompting#prompt-tuning) for the commands.

**Comment by dorost1234** [comment]
Hi
Thank you. I will run it for the seeds/train_seed/templates as you mentioned, could you confirm please if you considered results of all 80 cases and report min/mean or did you compute the average/min across templates for each seed/train_seed and then report the variance across seed/train_seed only? as I think the codes work like this. thanks

**Comment by shmsw25** [comment]
Good question. I reported mean and min of all 80 cases. You are right that currently the code prints out template-wise. I had another script that reads the log files to automatically compute the numbers to report in the paper which is not included in the repo right now.

**Comment by dorost1234** [comment]
Dear @shmsw25 
thank you for the explanation, now this is more clear for me. 
You mentioned, you choose the lr based on the trianing loss, I see in the codes this value is not saved. do you mind directing me please on which line I should compute this value? So there is a loss for each label, could you tell me please how you computed the training loss to choose the learning rate?
many thanks.

**Comment by shmsw25** [comment]
Training losses are recorded as a log file during training. Please refer to [this line](https://github.com/shmsw25/Channel-LM-Prompting/blob/master/run.py#L81).

**Comment by dorost1234** [comment]
thanks
